packages/dai-plugin-governance/src/SpellService.js

```diff
@@ -0,0 +1,94 @@
+import { PublicService } from '@makerdao/services-core';
+import { PAUSE } from './utils/constants';
+import DsSpellAbi from '../contracts/abis/DSSpell.json';
+import padStart from 'lodash/padStart';
+import assert from 'assert';
+import contractInfo from '../contracts/contract-info.json';
+const pauseInfo = contractInfo.pause;
+import { netIdToName } from './utils/helpers';
+
+export default class SpellService extends PublicService {
+  constructor(name = 'spell') {
+    super(name, ['smartContract', 'web3']);
+    this.eta = {};
+    this.done = {};
+    this.action = {};
+    this.executionDate = {};
+  }
+
+  getDelayInSeconds() {
+    if (this.delay) return this.delay;
+    this.delay = this._pauseContract().delay();
+    return this.delay;
+  }
+
+  async getEta(spellAddress) {
+    if (this.eta[spellAddress]) return this.eta[spellAddress];
+    const spell = this.get('smartContract').getContractByAddressAndAbi(
+      spellAddress,
+      DsSpellAbi
+    );
+    const eta = await spell.eta();
+    this.eta[spellAddress] = new Date(eta.toNumber() * 1000);
+    return this.eta[spellAddress];
+  }
+
+  async getScheduledDate(spellAddress) {
+    const delay = await this.getDelayInSeconds();
+    const eta = await this.getEta(spellAddress);
+    assert(eta, `spell ${spellAddress} has not been scheduled yet`);
+    return new Date(eta.getTime() - (delay * 1000));
+  }
+
+  async getDone(spellAddress) {
+    if (this.done[spellAddress]) return this.done[spellAddress];
+    const spell = this.get('smartContract').getContractByAddressAndAbi(
+      spellAddress,
+      DsSpellAbi
+    );
+    this.done[spellAddress] = spell.done();
+    return this.done[spellAddress];
+  }
+
+  async getAction(spellAddress) {
+    if (this.action[spellAddress]) return this.action[spellAddress];
+    const spell = this.get('smartContract').getContractByAddressAndAbi(
+      spellAddress,
+      DsSpellAbi
+    );
+    this.action[spellAddress] = spell.action();
+    return this.action[spellAddress];
+  }
+
+  async getExecutionDate(spellAddress) {
+    if (this.executionDate[spellAddress])
+      return this.executionDate[spellAddress];
```

Comment:
one thing i realize i missed here is assigning a value to `this.executionDate[spellAddress]` lower down in this function